### ISSUE TITLE
crypto module failure on small file

### ISSUE THREAD

**Issue by ibgreen** [visgl/loaders.gl#1033]
The CRC32 for the following file fails in loaders.gl/crypto:

The CRC32C hash I get by using loaders for this file is `+F10` while the hash computed by Google is `D4XXQQ==`.

Looks like we have a bug there that affects only a subset of files? Not sure if the file size (this one is ~1k) has to do with it somehow.

The CSV file in question:
```
placekey,date,cnt,hex,lat,lng
@5vg-7gw-wp9,2019-01-07 00:00,624,8a283082bdaffff,37.788316,-122.410113
@5vg-834-x3q,2019-01-07 00:00,1087,8a283095c027fff,37.733243,-122.490104
@5vg-7gt-w6k,2019-01-07 00:00,44,8a283082a80ffff,37.779138,-122.398532
@5vg-82h-kcq,2019-01-07 00:00,73,8a2830950637fff,37.72326,-122.435363
@5vg-7gv-66k,2019-01-07 00:00,343,8a283082ab1ffff,37.782134,-122.404228
@5vg-7gq-5pv,2019-01-07 00:00,119,8a283082800ffff,37.774349,-122.419324
@5vg-7gq-94v,2019-01-07 00:00,26,8a283082815ffff,37.772831,-122.421241
@5vg-7h2-t9z,2019-01-07 00:00,354,8a283082e7a7fff,37.769957,-122.386952
@5vg-7gw-w8v,2019-01-07 00:00,189,8a283082bd87fff,37.788659,-122.408628
@5vg-82h-bkz,2019-01-07 00:00,44,8a283095039ffff,37.721549,-122.442783
@5vg-82j-n5z,2019-01-07 00:00,179,8a283095119ffff,37.724993,-122.460535
@5vg-7gs-2p9,2019-01-07 00:00,64,8a2830829457fff,37.786862,-122.435116
@5vg-7gw-n5z,2019-01-07 00:00,66,8a283082bacffff,37.797449,-122.431227
@5vg-7zy-pvz,2019-01-07 00:00,188,8a28309466c7fff,37.701216,-122.484183
```

**Comment by belom88** [comment]
I put data in file:
[crc32.zip](https://github.com/visgl/loaders.gl/files/5671604/crc32.zip)

When tried to calculate hash with loaders, I got `ioBgAQ==`.
I made it using testcase:
```
test.only('CRC32CHashTransform#hashSync(CRC32, CSV, against external hash)', async t => {
  const response = await fetchFile(CSV_URL);
  const data = await response.arrayBuffer();

  const hash = CRC32CHashTransform.hashSync(data);
  t.equal(hash, CSV_CRC32, 'sync hash is correct');

  t.end();
});
```
How the hash "computed by Google" (`D4XXQQ==`) was calculated? It is not clear for me.

**Comment by belom88** [comment]
Further more:
I compared it with https://github.com/google/crc32c .
I've built this library and this code snippet
```
#include <iostream>
#include <fstream> 
#include "crc32c/crc32c.h"
using namespace std;

int main()
{
    string filename("/home/user/apps/tmp/crc32.csv");
    std::ifstream file(filename, std::ios::binary);
    
    file.seekg (0, file.end);
    int size = file.tellg();
    file.seekg (0, file.beg);

    std::uint8_t buffer[size];
    
    if (file.read((char*)&buffer, size))
    {
        std::uint32_t result;
        
        for (unsigned char sign: buffer) {
            cout << sign;
        }
        cout << endl;

        // Process a raw buffer.
        result = crc32c::Crc32c(buffer, size);

        cout << "Crc32c: " << std::hex <<  result << std::dec << endl;
        cout << "Size: " << size << endl;
    } else {
        cout << "file.read failed. Size: " << size  << endl;
    }
    
	return 0;
}
```
returns 0x8a806001 (`ioBgAQ==` in base64).
![image](https://user-images.githubusercontent.com/18549629/101759105-3ae86880-3aea-11eb-9609-3708c43dc9db.png)

So it looks like "loaders" crc32c transform works correctly.

**Comment by ibgreen** [comment]
I wonder if this happens when transforming in streaming mode using an async iterator

**Comment by belom88** [comment]
The same result in test case 'CRC32CHashTransform#iterator(CSV stream, against external hash)'
https://github.com/visgl/loaders.gl/blob/86cf7a1408e032f88180806c6a75f012d38767c5/modules/crypto/test/lib/crc32c-hash-transform.spec.js#L19

**Comment by kylebarron** [comment]
I was wondering originally whether it had to do with how the file was loaded. E.g. I'd gotten snagged in Node before loading a small file that had extra empty bytes at the end of the buffer because the default size for buffer chunks was larger than my file.

**Comment by manassra** [comment]
@belom88 to get the result of hashing a file from GCloud you can issue this command:
```
gsutil hash FootTraffic_short.csv
```
        Hash (crc32c):          D4XXQQ==
        Hash (md5):             vD4Rh9VRnJ0krRn1isK0mQ==
```

I've included the original file within a zip (github doesn't allow attaching csvs directly) in case you want the exact file.

**Comment by manassra** [comment]
The code using loaders that generated the wrong hash is run on the browser (not node), and uses `parseInBatches` with `CRC32CHashTransform` as a transform:

```js

  const batchIterator = await parseInBatches(source, STUDIO_LOADERS, {
    ...STUDIO_LOADER_OPTIONS,
    fetch: droppedFilesFS.fetch,
    metadata: true,
    // Note: inject streaming CRC32c hash calculation transform on input data
    transforms: [CRC32CHashTransform],
    crypto: {
      onEnd: result => {
        hash = result.hash;
      }
    }
  });
```

**Comment by belom88** [comment]
I created the PR. The root cause is incorrect conversion of hex string to base64 because of lack of leading zero.
'f85d741' (0f 85 d7 41) => 'D4XXQQ=='

**Comment by ibgreen** [comment]
Looks like we have a patch, good work. We need to land to master, cherry-pick to `2.3-release` and republish a 2.3.X patch (@vdunaev )

**Comment by manassra** [comment]
thank you @belom88!

**Comment by vdunaev** [comment]
Hi everyone! I've published 2.3.7 patch with this fix
